File: frontend/lib/features/challenges/data/models/requests/challenge_daily_tracking.dart
class ChallengeDailyTrackingsGetRequestModel {
  final List<String> challengeIds;

  const ChallengeDailyTrackingsGetRequestModel({required this.challengeIds});

  Map<String, dynamic> toJson() {
    return {
      'challenge_ids': challengeIds,
    };
  }
}

class ChallengeDailyTrackingUpdateRequestModel {
  final String habitId;
  final int dayOfProgram;
  final int quantityPerSet;
  final int quantityOfSet;
  final String unitId;
  final int weight;
  final String weightUnitId;
  final String? note;

  const ChallengeDailyTrackingUpdateRequestModel({
    required this.habitId,
    required this.dayOfProgram,
    required this.quantityPerSet,
    required this.quantityOfSet,
    required this.unitId,
    required this.weight,
    required this.weightUnitId,
    required this.note,
  });

  Map<String, dynamic> toJson() {
    return {
      'habit_id': habitId,
      'day_of_program': dayOfProgram,
      'quantity_per_set': quantityPerSet,
      'quantity_of_set': quantityOfSet,
      'unit_id': unitId,
      'weight': weight,
      'weight_unit_id': weightUnitId,
      'note': note,
    };
  }
}

class ChallengeDailyTrackingCreateRequestModel {
  final String challengeId;
  final String habitId;
  final int dayOfProgram;
  final int quantityPerSet;
  final int quantityOfSet;
  final String unitId;
  final int weight;
  final String weightUnitId;
  final int repeat;
  final String? note;

  const ChallengeDailyTrackingCreateRequestModel({
    required this.challengeId,
    required this.habitId,
    required this.dayOfProgram,
    required this.quantityPerSet,
    required this.quantityOfSet,
    required this.unitId,
    required this.weight,
    required this.weightUnitId,
    required this.repeat,
    required this.note,
  });

  Map<String, dynamic> toJson() {
    return {
      'challenge_id': challengeId,
      'habit_id': habitId,
      'day_of_program': dayOfProgram,
      'quantity_per_set': quantityPerSet,
      'quantity_of_set': quantityOfSet,
      'unit_id': unitId,
      'weight': weight,
      'weight_unit_id': weightUnitId,
      'repeat': repeat,
      'note': note,
    };
  }
}

File: backend/src/features/challenges/structs/requests/challenge_daily_tracking.rs
use std::collections::HashSet;

use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Deserialize)]
pub struct GetChallengeDailyTrackingParams {
    pub challenge_daily_tracking_id: Uuid,
}

#[derive(Deserialize)]
pub struct UpdateChallengeDailyTrackingParams {
    pub challenge_daily_tracking_id: Uuid,
}

#[derive(Deserialize)]
pub struct GetChallengeDailyTrackingsParams {
    pub challenge_id: Uuid,
}

#[derive(Deserialize)]
pub struct GetMultipleChallengesDailyTrackingsRequest {
    pub challenge_ids: Vec<Uuid>,
}

#[derive(Deserialize, Serialize)]
pub struct ChallengeDailyTrackingUpdateRequest {
    pub day_of_program: i32,
    pub habit_id: Uuid,
    pub quantity_per_set: f64,
    pub quantity_of_set: i32,
    pub unit_id: Uuid,
    pub weight: i32,
    pub weight_unit_id: Uuid,
    pub note: Option<String>,
    pub days_to_repeat_on: HashSet<i32>,
    pub order_in_day: i32,
}

#[derive(Deserialize, Serialize)]
pub struct ChallengeDailyTrackingCreateRequest {
    pub challenge_id: Uuid,
    pub habit_id: Uuid,
    pub day_of_program: i32,
    pub quantity_per_set: f64,
    pub quantity_of_set: i32,
    pub unit_id: Uuid,
    pub weight: i32,
    pub weight_unit_id: Uuid,
    pub days_to_repeat_on: HashSet<i32>,
    pub note: Option<String>,
    pub order_in_day: i32,
}
